client/client_test.go

```diff
@@ -33,6 +34,91 @@ func TestAddGetLeaf(t *testing.T) {
 	// TODO: Build a GetLeaf method and test a full get/set cycle.
 }
 
+func TestGetByIndex(t *testing.T) {
+	env, err := integration.NewLogEnv(context.Background(), 0, "TestGetByIndex")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer env.Close()
+	logID, err := env.CreateLog()
+	if err != nil {
+		t.Fatalf("Failed to create log: %v", err)
+	}
+
+	cli := trillian.NewTrillianLogClient(env.ClientConn)
+	client := New(logID, cli, testonly.Hasher, env.PublicKey)
+	// Add a few test leaves.
+	leafData := [][]byte{
+		[]byte("A"),
+		[]byte("B"),
+		[]byte("C"),
+	}
+
+	// TODO(gdbelvin): Replace with batch API.
+	// TODO(gdbelvin): Replace with AddSequenedLeaves API.
+	for _, l := range leafData {
+		ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(timeout))
+		defer cancel()
+		if err, want := client.AddLeaf(ctx, l), codes.DeadlineExceeded; grpc.Code(err) != want {
```

Comment:
A blocking AddLeaf is indeed bad, but if a client distrusts Trillian, how else should it be convinced that it's data was integrated? To not block, the client needs an SCT and the associated SCT verification routines. 

More concretely, when publishing a map head, the verifiable map personality's sequencer can't proceed until it can provide an inclusion proof for the map head back to clients.